lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -2523,59 +2549,107 @@ bool EliminateSliceConcat::run(Function *F, const CompilationContext &cctx) {
+      unsigned_t dim = -1;
       if (lastSN && (lastSN->getInput() != SN->getInput() ||
-                     !areSlicesConsecutive(lastSN, SN, CN->getDim()))) {
-        consecutiveSlices.emplace_back(currConsecutiveSlices);
+                     !findConsecutiveSliceDim(lastSN, SN, &dim))) {
+        consecutiveSlices.emplace_back(lastDim, currConsecutiveSlices);
         currConsecutiveSlices.clear();
       }
+      lastDim = dim;
       lastSN = SN;
       currConsecutiveSlices.emplace_back(SN);
     }
     if (currConsecutiveSlices.size()) {
-      consecutiveSlices.emplace_back(currConsecutiveSlices);
+      consecutiveSlices.emplace_back(lastDim, currConsecutiveSlices);
     }
 
     // Mapping from old Slices to new Slices where the range of each old Slice
     // is a subset of the corresponding new Slice
     std::unordered_map<SliceNode *, SliceNode *> oldToNewSlices;
-    for (const auto &slices : consecutiveSlices) {
+
+    // Mapping from new fused slices to reshape nodes that maybe be needed in
+    // case the fused slices need to be reshaped before concat. This happens
+    // when consecutive slices dimension differs from concat dimension by 1.
+    std::unordered_map<SliceNode *, ReshapeNode *> newSlicesToReshape;
+
+    for (const auto &slicePairs : consecutiveSlices) {
+      auto slicesDim = slicePairs.first;
+      auto slices = slicePairs.second;
+
       if (slices.size() <= 1) {
         continue;
       }
+      if (slicesDim != CN->getDim() &&
+          slicesDim != CN->getDim() + 1 &&
+          slicesDim != CN->getDim() - 1) {
+        // Optimizations are possible only if:
+        // 1) slices consecutive dimension is the same as concat dimension, or
+        // 2) slices consecutive dimension is adjact to the concat dimension.
+        //
+        // In the former case, we replace it with a single slice. In the latter
+        // case, we replace it with a single slice and reshape.
+        continue;
+      }
+
       SliceNode *firstSlice = slices.front();
       auto *srcNode = firstSlice->getInput().getNode();
       std::vector<dim_t> endDims;
+
       for (size_t i = 0, e2 = firstSlice->getResult().dims().size(); i < e2;
            i++) {
         endDims.emplace_back(slices.back()->getStart()[i] +
                              slices.back()->getResult().dims()[i]);
       }
       auto *newSlice = F->createSlice(firstSlice->getName(), srcNode,
                                       firstSlice->getStart(), endDims);
+
       for (auto *slice : slices) {
         oldToNewSlices[slice] = newSlice;
       }
+
+      // Create a reshape node based on consecutive slice dimension and
+      // concat dimension.
+      if (slicesDim == CN->getDim() + 1 ||
+          slicesDim == CN->getDim() - 1) {
+        auto outputDimVec = newSlice->getResult().dims().vec();
+        outputDimVec[CN->getDim()] *= outputDimVec[slicesDim];
+        outputDimVec[slicesDim] = 1;
```

Comment:
Thought about this logic, and did a couple of experiment in Numpy, and it seems to hold. Still 100% sure, so some extra attention during the review may be warranted.

Reply:
Yeah, I think this is wrong. As an example, in your unit test, change the first few lines to the following, and it will fail:

```cpp
  auto *src =
      mod_.createPlaceholder(ElemKind::FloatTy, {8, 5, 3}, "src", false);
  auto *A = F_->createSlice("A", src, {0, 0, 0}, {2, 5, 3});
  auto *B = F_->createSlice("B", src, {2, 0, 0}, {4, 5, 3});
  auto *C = F_->createSlice("C", src, {4, 0, 0}, {6, 5, 3});
```

I think you want to use the original dimension of one of the slices.

Reply:
The optimization will only work if the size of the slice along the consecutive dimension is 1. This did bring up another problem, that I was not checking if this was the case. A check, along with a comment, has been added. So in your example, this optimization will do nothing.